jobserver/api.py

```diff
@@ -273,3 +276,45 @@ def get(self, request, *args, **kwargs):
 
         actions_with_status = workspace.get_action_status_lut()
         return Response(actions_with_status, status=200)
+
+
+class ReleaseUploadAPI(APIView):
+    # DRF file upload does not use multipart, is just a simple byte stream
+    parser_classes = [FileUploadParser]
+
+    def post(self, request, workspace_name):
+        try:
+            workspace = Workspace.objects.get(name=workspace_name)
+        except Workspace.DoesNotExist:
+            return Response(status=404)
+
+        # authenticate and get backend
+        backend = get_backend_from_token(request.headers.get("Authorization"))
+        backend_user = request.headers.get("Backend-User", "").strip()
+        if not backend_user:
+            raise NotAuthenticated("Backend-User not valid")
+
+        # TODO: validate user once we have mappings from backend username.
+
+        release_hash = request.headers.get("If-None-Match")
```

Comment:
I've not used this header before, [MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-None-Match) suggests a non GET/HEAD requests are related to saving files, could you document in a comment here why we're using the header in this instance.

Reply:
Well, we are saving a file :smiley:  - but good idea to add docs

Reply:
Putting a hash (ie etag) in If-None-Match in a POST is not an explicitly documented usage, but is arguably valid and I've done it before, but its supposed to be optional, and I was requiring it, which made it all awkward.

Then I realised I can just use PUT with the hash in the url, and it's all simple much simpler, so I did that. Only reason I didn't do that originally was because I started out with the client not knowing the id, which meant a POST made sense. Half way through, I switched to using the hash as id as a means to detect republishing the same release, and missed that PUT with a full url was better.